### PR TITLE
Refactor creator

### PR DESCRIPTION
**Pull request by ReyRizki**

### CODE REVIEW

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java [hunk 1]
```diff
@@ -27,81 +27,64 @@ public Trie createTranssactionTrie() throws IOException {
...
+        int productsLength = products.length;
+
+        for (int i = 0; i < productsLength; i++) {
+            Node currentNode = trie.getNodeAt(trav);
 
-                if (currentNode.isLeaf()) {
+            int lowerBound = lowerBoundIndex(trie, currentNode, products[i]);
+            Node currentNodeChild = trie.getNodeAt(currentNode.getChildIndexAt(lowerBound));
+
+            if (currentNodeChild.isNull() || !currentNodeChild.getProduct().equals(products[i])) {
```
Comment (fauh45) [inline src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:50]: Better put it into a boolean variable
```suggestion
            boolean currentNodeProductisNotSame = !currentNodeChild.getProduct().equals(products[i]);
            boolean currentNodeisNotExpected = currentNodeChild.isNull() || currentNodeProductisNotSame;
            if (currentNodeisNotExpected) {
```

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java [hunk 2]
```diff
@@ -118,4 +101,17 @@ private void shiftIndexes(Trie trie, int target) {
             }
         }
     }
+
+    private int lowerBoundIndex(Trie trie, Node node, String product) {
+        // traverse the child indexes
+        int j = 0;
+        Node nodeChild = trie.getNodeAt(node.getChildIndexAt(j));
+
+        while (!nodeChild.isNull() && nodeChild.getProduct().compareTo(product) < 0 && j < node.getDegree()) {
```
Comment (fauh45) [inline src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:110]: Conditional too long?

File: src/main/java/org/pawaneuler/DataTypes/Trie/Trie.java [hunk 3]
```diff
@@ -55,7 +55,7 @@ public void addNode(int index, Node node) {
     public Node getNodeAt(int index) {
         if (index >= 0 && index < nodes.size())
             return this.nodes.get(index);
-        else
+        else if (index < 0 || index > nodes.size());
             return Node.getNullNode();
```
Comment (fauh45) [inline src/main/java/org/pawaneuler/DataTypes/Trie/Trie.java]: Better to make it into default return?

```suggestion
    return Node.getNullNode();
```

File: src/main/java/org/pawaneuler/DataTypes/Trie/Trie.java [hunk 4]
```diff
@@ -55,7 +55,7 @@ public void addNode(int index, Node node) {
     public Node getNodeAt(int index) {
         if (index >= 0 && index < nodes.size())
             return this.nodes.get(index);
-        else
+        else if (index < 0 || index > nodes.size());
```
Comment (ReyRizki) [inline src/main/java/org/pawaneuler/DataTypes/Trie/Trie.java]:
```suggestion
        else
```

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java [hunk 5]
```diff
@@ -118,4 +101,17 @@ private void shiftIndexes(Trie trie, int target) {
             }
         }
     }
+
+    private int lowerBoundIndex(Trie trie, Node node, String product) {
+        // traverse the child indexes
+        int j = 0;
```
Comment (fauh45) [inline src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:107]: Non-descriptive name

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java [hunk 6]
```diff
@@ -27,81 +27,64 @@ public Trie createTranssactionTrie() throws IOException {
...
+            tcsvData = source.readLine();
+            products = tcsvData.getProducts();
+        }
+
+        return trie;
+    }
+
+    private void insertProduct(Trie trie, String[] products) {
+        // traverse the trie from the root node
+        int trav = 0;
```
Comment (fauh45) [inline src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:41]: Non-descriptive name, use `currentTraverseIndex` instead?

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java [hunk 7]
```diff
@@ -27,81 +27,64 @@ public Trie createTranssactionTrie() throws IOException {
...
+        int productsLength = products.length;
+
+        for (int i = 0; i < productsLength; i++) {
```
Comment (fauh45) [inline src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:44]: Non descriptive name of `i`
```suggestion
        for (int currentProductIndex = 0; i < productsLength; currentProductIndex++) {
```

**Review by fauh45 — CHANGES_REQUESTED** [review]
Refactoring not complete

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 2 above)
Comment (ReyRizki) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:110]: I can't decide the variable name. Any suggestion?

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 5 above)
Comment (ReyRizki) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:107]: what name should I use, then?

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 1 above)
Comment (ReyRizki) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:50]: yes, Sir

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 6 above)
Comment (ReyRizki) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:41]: umm okay

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 7 above)
Comment (ReyRizki) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:44]: I think i is okay

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 2 above)
Comment (fauh45) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:110]:
```suggestion
        boolean nodeIsInsideLimit = j < node.getDegree();
        boolean nodeIsBeforeCurrentProduct = nodeChild.getProduct().compareTo(product) < 0;
        boolean nodeIsNotLowerBound = !nodeChild.isNull() && nodeIsBeforeCurrentProduct && nodeIsInsideLimit;
        while (nodeIsNotLowerBound) {
```

Maybe will work?

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 5 above)
Comment (fauh45) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:107]: `currentChildIndex` is it better?

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 7 above)
Comment (fauh45) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:44]: A bit too vague, but heh, I'll pass it lol

File: src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java (hunk 5 above)
Comment (ReyRizki) [reply src/main/java/org/pawaneuler/Generator/TransactionTrieGenerator/TransactionTrieCreator.java:107]: umm okay

**Pull request merged**